File: erettsegi/kozep/2024_may/java/Letra.java
import module java.base;

public class Letra {

    public static void main(String[] args) throws Exception {
        var dobasok = Files.readString(Path.of("dobasok.txt")).trim().split(", ");
        var lepesek = new ArrayList<LepesAllapot>();

        for(var dobas : dobasok) {
            var dobasErtek = Integer.parseInt(dobas);
            var elozoLepes = lepesek.isEmpty() ? new LepesAllapot(0, 0, false) : lepesek.getLast();

            lepesek.add(lepestAlkalmaz(elozoLepes, dobasErtek));
        }

        System.out.print("2. Feladat: ");

        for(var lepes : lepesek) {
            System.out.print(lepes.ertek + " ");
        }

        System.out.println();

        var visszalepesekSzama = 0;
        for(var lepes : lepesek) {
            if(lepes.visszalepes) {
                ++visszalepesekSzama;
            }
        }

        System.out.println("3. Feladat: A játék során " + visszalepesekSzama + " alkalommal lépett létrára");
        System.out.println("4. Feladat: " + (lepesek.getLast().ertek > 45 ? "A játékot befejezte" : "A játékot abbahagyta"));
    }

    static LepesAllapot lepestAlkalmaz(LepesAllapot elozoAllapot, int dobottErtek) {
        var osszeg = elozoAllapot.ertek + dobottErtek;
        var visszalepes = osszeg % 10 == 0;

        return new LepesAllapot(elozoAllapot.dobasI + 1, visszalepes ? osszeg - 3 : osszeg, visszalepes);
    }

    record LepesAllapot(int dobasI, int ertek, boolean visszalepes) {}
}

File: erettsegi/kozep/2024_may/python/letra.py
from dataclasses import dataclass

@dataclass
class LepesAllapot:
    dobas_i: int
    ertek: int
    visszalepes: bool

with open('dobasok.txt') as dobasok_input:
    dobasok = [ int(k) for k in dobasok_input.readline().strip().split(', ') ]

lepesek = []
for dobas in dobasok:
    elozo_lepes = LepesAllapot(0, 0, False) if len(lepesek) == 0 else lepesek[-1]
    osszeg = elozo_lepes.ertek + dobas
    visszalepes = osszeg % 10 == 0

    lepesek.append(LepesAllapot(elozo_lepes.dobas_i + 1, osszeg - 3 if visszalepes else osszeg, visszalepes))

print('2. Feladat: ')

for lepes in lepesek:
    print(f'{lepes.ertek} ', end = '')

print()

visszalepesek_szama = sum(1 if k.visszalepes else 0 for k in lepesek)

print(f'3. Feladat: A játék során {visszalepesek_szama} alkalommal lépett létrára')
print(f'4. Feladat: {"A játékot befejezte" if lepesek[-1].ertek > 45 else "A játékot abbahagyta"}')
